### PR TITLE
Self-host the full Stardag stack on Modal: `stardag self-host up` (+ local email/password auth mode)

### PR DESCRIPTION
**Pull request by andhus**
Implements #187.

## What

One command brings up a private Stardag service (Registry API + web UI) on Modal with Postgres on Neon; one command updates it:

```bash
uvx --from "stardag[selfhost]" stardag self-host up       # provision DB, migrate, deploy, print UI URL
uvx --from "stardag[selfhost]" stardag self-host upgrade  # migrate + redeploy from current source
```

Docs: `docs/how-to/self-host-modal.md` (quickstart + step-by-step account setup for both auth modes).

## Commits (reviewable independently)

1. **Runtime auth config** — `GET /api/v1/auth/config` now serves the full client auth configuration; the UI resolves it at boot instead of requiring build-time `VITE_*` baking (which still takes precedence for backwards compatibility).
2. **Serverless/pooler hardening** — `STARDAG_API_DATABASE_URL_DIRECT` (migrations bypass transaction-mode poolers) and `STARDAG_API_DATABASE_POOLER_COMPAT` (SQLAlchemy-documented asyncpg/PgBouncer recipe); warning when the internal JWT keypair is ephemeral.
3. **Local auth mode** (`AUTH_MODE=local`) — email/password accounts managed by the API itself, minting user-scoped session tokens (`token_use=session`, strictly separated from workspace tokens) that plug into the existing exchange flow. Login/register/change-password endpoints (registration off by default), bcrypt with timing-uniform verification, login rate limiting, idempotent bootstrap-admin provisioning at startup. UI sign-in form + AuthContext branch; `stardag auth login` auto-detects the mode.
4. **`stardag self-host` CLI + Modal server app** (extra: `stardag[selfhost]`) — Neon provisioning from an API key (find-or-create project, Postgres 16), single Modal ASGI endpoint serving API + UI same-origin with SPA fallback, one-off migrate function, JWT keypair secret created once and preserved across upgrades. The UI is npm-built inside the Modal image build, so users need no local Node/Docker.
5. **Docs** + skill-bundle updates.

## Verification

- 335 API + 920 SDK + 84 UI tests pass (new coverage for every added surface).
- Live smoke deployment on Modal: image build (in-image UI compile), `/health`, runtime auth config, static UI serving, SPA fallback, and JWKS all verified working; a full E2E against a real Neon project is in progress and will be reported here before undrafting.
- OIDC mode (the default) is behaviorally unchanged; the existing AWS CDK deployment path is unaffected.

## Out of scope (follow-ups)

- Prebuilt server images published per release (from-source build works today; this is a speed optimization).
- UI for password change in local mode (endpoint exists).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01ARrhAGii9mMcpyVNTbSMEf

### CODE REVIEW

**Comment by andhus** [comment]
Full end-to-end verified against a real Neon project and Modal workspace:

- `self-host up --yes` provisioned the Neon project (Postgres 16) from an org API key, applied all 15 migrations via the direct endpoint, and deployed in ~18s (image cached from a prior build; cold build is ~80s).
- Live checks against the deployed URL: `/health`, `/api/v1/auth/config` (local mode), bootstrap-admin login → session token → `/ui/me` (personal workspace auto-created) → workspace token exchange, and the UI served same-origin.

Marking ready for review.

File: lib/stardag/src/stardag/_cli/selfhost.py [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+        pooler_compat = False
+    else:
+        pooled_url, direct_url, pooler_compat = _resolve_database_urls(
+            database_url,
+            database_url_direct,
+            neon_api_key,
+            neon_project,
+            interactive,
+            neon_pg_version=neon_pg_version,
+        )
```
[the remark this answers is not in the document]
Comment (andhus) [reply lib/stardag/src/stardag/_cli/selfhost.py:492]: Fixed in 8e3d7b7: `self-host up` now fails fast when any auth/config flag (`--auth-mode`, `--admin-email`, `--admin-password`, `--enable-registration`, `--oidc-*`) is provided while the config secret already exists and no database inputs are given, with an error explaining that reconfiguration requires re-supplying `--neon-api-key` or `--database-url`. Added a unit test for the flag-detection helper.

File: app/stardag-api/src/stardag_api/auth/dependencies.py [hunk 2]
```diff
@@ -638,7 +638,21 @@ async def get_current_user_flexible(
...
+        user = await get_user_by_id(db, session_payload.sub)
+        if user is None:
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+                detail="User not found",
+            )
+        return user
+    except (TokenExpiredError, TokenInvalidError):
+        # Not a valid session token, try OIDC
         pass
```
[the remark this answers is not in the document]
Comment (andhus) [reply app/stardag-api/src/stardag_api/auth/dependencies.py]: Fixed in 8e3d7b7: `get_current_user_flexible` now only attempts session-token validation when `auth_settings.mode == "local"`, so session tokens minted before a local→OIDC switch (with unchanged JWT keys) can no longer authenticate until expiry. Added a regression test asserting a local-mode session token gets 401 on `/ui/me` after switching the mode to `oidc`.

File: app/stardag-api/src/stardag_api/routes/auth.py [hunk 3]
```diff
@@ -124,44 +188,95 @@ async def get_jwks_endpoint():
...
+    """
+    token_str = credentials.credentials
+
+    token_manager = get_token_manager()
+    try:
+        session_payload = token_manager.validate_session_token(token_str)
+    except TokenError:
+        session_payload = None
+
+    if session_payload is not None:
```
[the remark this answers is not in the document]
Comment (andhus) [reply app/stardag-api/src/stardag_api/routes/auth.py:238]: Fixed in 8e3d7b7: `get_exchange_user` now only tries session-token validation when `auth_settings.mode == "local"`; in OIDC mode the credential must be a valid OIDC token. Covered by the same regression test (session token gets 401 on `/auth/exchange` after switching to `oidc`).

File: app/stardag-api/src/stardag_api/services/local_auth.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    result = await db.execute(
+        select(User).where(func.lower(User.email) == email.lower())
+    )
+    user = result.scalar_one_or_none()
+    if user is None:
+        await create_local_user(db, email=email, password=password)
+        logger.info("Bootstrap admin created: %s", email)
+    elif user.password_hash is None:
+        user.password_hash = await hash_password_async(password)
+        await db.commit()
```
[the remark this answers is not in the document]
Comment (andhus) [reply app/stardag-api/src/stardag_api/services/local_auth.py:188]: Fixed in 8e3d7b7: `ensure_bootstrap_admin` now calls `validate_password_policy` before hashing/setting the bootstrap password on an existing passwordless user, matching `/auth/register` (the create-new-user path already validated via `create_local_user`). This also prevents silent bcrypt 72-byte truncation. Added a test.

File: app/stardag-ui/src/auth/config.ts [hunk 5]
```diff
@@ -1,37 +1,121 @@
...
-// Format: {domain-prefix}.auth.{region}.amazoncognito.com
-export const COGNITO_DOMAIN = import.meta.env.VITE_COGNITO_DOMAIN || "";
+// Auth configuration, resolved at app boot.
+//
+// Sources, in order of precedence:
+// 1. Build-time environment variables (VITE_OIDC_*) — kept for backwards
+//    compatibility with deployments that bake config into the bundle.
+// 2. Runtime config fetched from the API (GET /api/v1/auth/config) — allows
+//    a prebuilt UI bundle to be pointed at any IdP (or local auth mode)
+//    without rebuilding.
```
[the remark this answers is not in the document]
Comment (andhus) [reply app/stardag-ui/src/auth/config.ts]: Fixed in 8e3d7b7 (comment only). The behavior is intended: a server in local mode cannot exchange OIDC tokens, so a build-time issuer would just produce a broken login flow — the server-declared mode has to win. Reworded the header to state that the auth *mode* is server-declared, while the OIDC field values themselves prefer build-time `VITE_OIDC_*` settings field by field.

File: app/stardag-ui/src/auth/config.ts [hunk 6]
```diff
@@ -1,37 +1,121 @@
...
+      signal: AbortSignal.timeout(5000),
+    });
+    if (response.ok) {
+      runtime = (await response.json()) as RuntimeAuthConfigResponse;
+    } else {
+      console.warn(`[auth] /auth/config returned ${response.status}`);
+    }
+  } catch (error) {
+    console.warn("[auth] Failed to fetch runtime auth config:", error);
+  }
```
[the remark this answers is not in the document]
Comment (andhus) [reply app/stardag-ui/src/auth/config.ts:94]: Applied in 8e3d7b7: switched to `AbortController` + `setTimeout` (cleared in `finally`). Minor correction to the premise: the `AbortSignal.timeout()` call was evaluated inside the `try`, so on unsupported browsers the throw was caught and the UI fell back to build-time config rather than crashing — but for a self-hosted UI with no build-time config that fallback means `authMode="disabled"`, so avoiding the compatibility edge is still worthwhile.

**Review by andhus — COMMENTED** [review]
Reviewed the full branch HEAD (API local-auth mode, runtime UI auth config, self-host CLI + Modal app, Neon client, UI changes, docs). Overall this is a well-built PR: the session-token / workspace-token separation is enforced in both directions via the `token_use` claim, timing-uniform password verification (dummy bcrypt hash for unknown emails) is done correctly, registration is off by default, the bootstrap admin is idempotent and never overwrites an existing password, the JWT keypair secret is create-once, and test coverage of the new surfaces is genuinely thorough (including regression tests for the earlier review round's mode-switch issues). The findings below are hardening and edge-case items, not blockers.

**Finding 1 (severity: medium)** — `app/stardag-api/src/stardag_api/routes/auth.py:316-321` — `_client_ip()` trusts the *first* entry of `X-Forwarded-For`, which is client-supplied even behind a proxy that appends the real client IP (ALB and most reverse proxies append, so `split(",")[0]` is attacker-controlled). Since the login rate-limit key is `email|ip`, an attacker can rotate a spoofed XFF value to get a fresh bucket per request and brute-force a known email without ever hitting the 429 (bcrypt cost then being the only throttle). Suggested fix: take the *last* XFF entry (nearest trusted hop), or make trusted-proxy handling explicit/configurable; alternatively add a secondary per-email-only limiter (higher threshold) that ignores IP so header spoofing can't fully bypass throttling.

**Finding 2 (severity: low)** — `app/stardag-api/src/stardag_api/services/local_auth.py:39-51` — `LoginRateLimiter._attempts` is a `defaultdict(deque)` that only ever removes entries on successful login (`reset`); expired windows leave empty deques behind and every unique `email|ip` key (unauthenticated, attacker-chosen) adds a permanent dict entry. Combined with Finding 1 this is an unbounded-memory growth vector on a long-lived container. Suggested fix: delete the key in `check()` when its deque empties after pruning, and/or cap total tracked keys (LRU).

**Finding 3 (severity: low)** — `app/stardag-api/src/stardag_api/routes/auth.py` (`local_change_password`) — changing the password does not invalidate outstanding session tokens: they are stateless JWTs with a default TTL of 7 days (`session_token_ttl_hours = 24*7`), so a stolen session token survives a password change/rotation. Acceptable for the self-host target, but worth either (a) a claim-based cutoff (reject session tokens with `iat` older than a `password_changed_at` column) or (b) an explicit note in the self-host docs that password change does not log out existing sessions.

**Finding 4 (severity: low)** — `lib/stardag/src/stardag/_cli/selfhost.py:482` — the CLI validates only the 8-char minimum for the admin password, but the server also enforces a 72-byte maximum (`MAX_PASSWORD_LENGTH`, bcrypt truncation guard). A >72-byte `--admin-password` passes `self-host up`, gets written into the config secret, and then `ensure_bootstrap_admin` raises `PasswordPolicyError` in the FastAPI lifespan on every container start — a confusing post-deploy crash loop. Suggested fix: mirror the max-length (byte) check in the CLI prompt/validation.

**Finding 5 (severity: low)** — `lib/stardag/src/stardag/selfhost/_modal_app.py:122-131` — the SPA fallback serves `index.html` with HTTP 200 for *any* unmatched path, including unmatched `/api/v1/*` paths (routes win only when they exist; a typo'd or version-skewed API path falls through to the mount). SDK/API clients then get a 200 with an HTML body instead of a JSON 404, which surfaces as opaque decode errors. Suggested fix: in `SPAStaticFiles.get_response`, keep the 404 for paths starting with `api/` (and `health`, `.well-known/`).

**Finding 6 (severity: low)** — `lib/stardag/src/stardag/_cli/selfhost.py:564-570` — `self-host upgrade` takes `--keep-warm` with default 0 and always passes it through, so an operator who deployed with `up --keep-warm 1` and later runs a plain `upgrade` silently reverts to scale-to-zero. Suggested fix: persist the keep-warm choice (e.g. in the config secret) and default `upgrade` to the stored value, or at least call out the reset in the option help / docs.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01ARrhAGii9mMcpyVNTbSMEf

**Comment by andhus** [comment]
## Review findings addressed

All six findings from the review are resolved in 44039b2 (API, findings 1–3) and b30b227 (self-host CLI/Modal app, findings 4–6).

| # | Finding | Resolution |
|---|---------|------------|
| 1 (medium) | X-Forwarded-For spoofing defeats login rate limiting | Both suggested fixes applied in 44039b2: `_client_ip()` now takes the **last** XFF entry (appended by the nearest trusted hop) instead of the client-controlled first one, and a second per-email-only limiter (30 attempts / 15 min) runs alongside the existing `email\|ip` one, so IP rotation alone can't brute-force an account. Both limiters reset on successful login. Tests: spoofed XFF prefix rotation still trips the per-IP bucket; full IP rotation trips the per-email limiter; success resets both. |
| 2 (low) | Rate-limiter memory growth | 44039b2: keys are dropped as soon as their window fully expires, and once the tracked-key count exceeds a threshold (10k) each check first sweeps out all fully-expired entries. Test covers the sweep and key cleanup. |
| 3 (low) | Password change doesn't invalidate outstanding session tokens | Implemented the claim-based cutoff in 44039b2: new nullable timezone-aware `users.password_changed_at` column (autogenerated migration, up/down/up cycle verified), set in `local_change_password`; session tokens with `iat` before it are rejected 401 on both session-token surfaces (`get_current_user_flexible` session branch and `get_exchange_user`). Workspace access tokens are deliberately unaffected (10-min TTL). Tests: pre-change session token rejected on `/ui/me` and `/auth/exchange`, post-change token works, pre-change workspace token unaffected, helper handles unset/naive timestamps. |
| 4 (low) | CLI missing 72-byte max password check | b30b227: the CLI now mirrors the server's policy (`_admin_password_error`: 8-char min + 72-byte UTF-8 max), failing `self-host up` up front instead of crash-looping the container at startup. Unit tests include multi-byte UTF-8 cases. |
| 5 (low) | SPA fallback swallows unknown API paths | b30b227: the fallback in `_modal_app.py` now only serves `index.html` when the (mount-relative, no leading slash) path does not start with `api/`, `health`, or `.well-known/` — unknown API paths return a proper JSON 404. Mirrored with identical guard logic into `SPAStaticFiles` in `stardag_api/server.py` on the `server-release-pipeline` branch (#191) in 42b8562, with a unit test asserting `GET /api/v1/nonexistent` returns JSON 404, not HTML 200; the two copies converge cleanly when #191 merges. |
| 6 (low) | `upgrade --keep-warm` silently resets keep-warm to 0 | b30b227: the value is persisted in a `modal.Dict` named `<app>-meta`. `up`/`upgrade` write it only when the flag is explicitly provided (defaults changed to `None` so "not provided" is distinguishable) and otherwise reuse the stored value (default 0). `status` shows it; `destroy --delete-secrets` also deletes the Dict (`modal.Dict.objects.delete`, verified against modal 1.5). Resolution logic unit-tested with a mocked Dict; `--keep-warm` help text and the self-host docs updated. |

Verification: `app/stardag-api` 344 passed (incl. new tests), migration upgrade→downgrade→upgrade cycle green against local Postgres; `lib/stardag` `tests/test_selfhost` 14 passed (no skips, selfhost+modal extras installed); pre-commit (ruff, pyright, prettier) green on all touched files; on `server-release-pipeline` 342 passed incl. the new SPA-fallback test.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01ARrhAGii9mMcpyVNTbSMEf
